scripts/diff-pkg-changes.py

```diff
@@ -0,0 +1,291 @@
+            pkg_dict['Architecture'],
+            pkg_dict['Maintainer'],
+            pkg_dict['MD5Sum'],
+            pkg_dict['SHA256sum'],
+            int(pkg_dict['Size']),
+            pkg_dict['Filename'],
+            pkg_dict['Source'],
+            pkg_dict['Description'],
+            pkg_dict['OE'],
+            pkg_dict.get('HomePage', ''),
+            pkg_dict['License'],
+            pkg_dict['Priority']
+        ))
+
+    return pkgs
+
+# Create a final report from all the package versions
+def gen_report(
+    old_core_pkgs: list['Package'], new_core_pkgs: list['Package'],
+    old_extra_pkgs: list['Package'], new_extra_pkgs: list['Package'],
+    old_vers: str, new_vers: str
+) -> str:
+    report: str = \
+        '## Changes from Version ' \
+            + old_vers + ' to Version ' + new_vers + '\n\n### Changes to core/\n\n'
+
+    # Handle core diffs
+    for pkg in old_core_pkgs:
+        other_vers_of_pkg: list['Package'] = \
+            [ i for i, other_pkg in enumerate(new_core_pkgs) if other_pkg.name == pkg.name ]
+        if len(other_vers_of_pkg) == 0: # Check for package removal from version to version
+            report += '- Package \'' + pkg.name + '\' removed\n'
+        else:
+            other_pkg: 'Package' = new_core_pkgs[other_vers_of_pkg[0]] # We know there's only one FYI
+            diff_str: str = pkg.diff(other_pkg)
+            if diff_str != '':
+                report += diff_str
+            new_core_pkgs.pop(other_vers_of_pkg[0])
+    for pkg in new_core_pkgs:
+        # We removed packages that were duplicates, so what's left is new packages
+        report += '- Added package \'' + pkg.name + '\'\n'
+    report += '\n### Changes to extra/\n\n'
+
+    # Do the same for extra
+    for pkg in old_extra_pkgs:
+        other_vers_of_pkg: list['Package'] = \
+            [ i for i, other_pkg in enumerate(new_extra_pkgs) if other_pkg.name == pkg.name ]
+        if len(other_vers_of_pkg) == 0: # Check for package removal from version to version
+            report += '- Package \'' + pkg.name + '\' removed\n'
+        else:
+            other_pkg: 'Package' = new_core_pkgs[other_vers_of_pkg[0]] # We know there's only one FYI
+            diff_str: str = pkg.diff(other_pkg)
+            if diff_str != '':
+                report += diff_str
+            new_extra_pkgs.pop(other_vers_of_pkg[0])
+    for pkg in new_extra_pkgs:
+        # We removed packages that were duplicates, so what's left is new packages
+        report += '- Added package \'' + pkg.name + '\'\n'
+    report += '\n'
+
+    return report
+
+# Update the FEED_CHANGELOG with the new version info
+def update_changelog(report: str) -> None:
+    old_log: str = ''
+    with open('FEED_CHANGELOG.md', 'r') as changelog:
+        old_log = changelog.read()
+    new_log = FEED_CHANGELOG_START + report + old_log.replace(FEED_CHANGELOG_START, '')
+    with open('FEED_CHANGELOG.md', 'w') as changelog:
+        changelog.write(new_log)
```

Comment:
What do you think about using a more UNIX-y output design? I'm thinking that we could remove the hardcoded `FEED_CHANGELOG.md` paths, and instead have this script write the `report` as a JSON structure to stdout. Then we can keep a log of all the changes in the nilrt.git source like
```bash
python3 diff-pkg-changes.py ... >>./docs/feed-changelog.json
```

Then we can have a small companion script that reads json from stdin and writes out the markdown changelog like
```bash
python3 emit-feed-changelog-markdown.py <./docs/feed-changelog.json >./docs/feed-changelog.md
```

I would use `jinja2` templating for generating the markdown, but an implementation similar to what you are already doing in this script is probably fine too.

What do you think?

Reply:
I like this, but I have one question about the JSON: are we keeping just the most recent changes, or should it contain multiple sets of changes over time?

Because if we're updating the JSON overtime to keep changes every new version like the changelog, simply updating it with append (>>) won't work because when importing the JSON, python expects:
```
{
  "something": {...},
  "something_else": {...}
}
```
but append would create this:
```
{
  "something": {...}
}
{
  "something_else": {...}
}
```

There's a few ways we could go about creating the valid JSON like having an input/output file as an argument or modifying the output before appending in bash, what would be the preference here?

Reply:
We can just output the whole JSON object without newlines and then use the [jsonlines](https://jsonlines.org/) standard to store it. The changelog digest script would then just iterate through each line of the file and load it as a json object.